addons/mass_mailing/models/mailing_contact.py

```diff
@@ -38,6 +47,36 @@ def create(self, vals_list):
         return super().create(vals_list)
 
     def write(self, vals):
+        for record in self:
+            if 'opt_out' in vals or 'opt_out_reason' in vals:
+                vals['opt_out_reason'] = vals.get('opt_out', record.opt_out) and vals.get('opt_out_reason')
+            if 'opt_out' in vals or 'opt_out_custom' in vals:
+                vals['opt_out_custom'] = vals.get('opt_out', record.opt_out) and vals.get('opt_out_custom')
+            author_id = self.env.user.partner_id.id
+            if 'opt_out' in vals:
+                message = _('The recipient <strong>unsubscribed from %s</strong> mailing list') \
+                        if vals['opt_out'] else _('The recipient <strong>subscribed to %s</strong> mailing list')
+                record.contact_id.sudo().message_post(
+                    body=message % record.list_id.name,
+                    author_id=author_id
+                )
+            if vals.get('opt_out_reason'):
+                if vals['opt_out_reason'] == 'other':
+                    record.contact_id.sudo().message_post(
+                        body=_("Feedback from %s: %s (other)") % (
+                            record.contact_id.email,
+                            html_escape(vals['opt_out_custom'])
+                        ),
+                        author_id=author_id
+                    )
+                else:
+                    record.contact_id.sudo().message_post(
```

Comment:
Here you are adding a way to post on records using sudo. This is a security issue, caller should check access is granted and call write as sudo.